src/lib/polygon/PolygonKey.js

```diff
@@ -0,0 +1,145 @@
+/* global Nimiq */
+/* global Errors */
+/* global ethers */
+/* global CONFIG */
+
+class PolygonKey { // eslint-disable-line no-unused-vars
+    /**
+     * @param {Key} key
+     */
+    constructor(key) {
+        if (key.type !== Nimiq.Secret.Type.ENTROPY) {
+            throw new Errors.KeyguardError('Key must be of type Nimiq.Entropy');
+        }
+        this._key = key;
+    }
+
+    /**
+     * @param {string} path
+     * @returns {string}
+     */
+    deriveAddress(path) {
+        const wallet = this.deriveKeyPair(path);
+        return wallet.address;
+    }
+
+    /**
+     * @param {string} path
+     * @param {ethers.providers.TransactionRequest} transaction
+     * @returns {Promise<string>}
+     */
+    async sign(path, transaction) {
+        const wallet = this.deriveKeyPair(path);
+        return wallet.signTransaction(transaction);
+    }
+
+    /**
+     *
+     * @param {string} path
+     * @param {ethers.TypedDataDomain} domain
+     * @param {Record<string, Array<ethers.TypedDataField>>} types
+     * @param {Record<string, any>} value
+     * @returns {Promise<string>}
+     */
+    async signTypedData(path, domain, types, value) {
+        const wallet = this.deriveKeyPair(path);
+        return wallet._signTypedData(domain, types, value);
+    }
+
+    /**
+     * @param {string} path
+     * @param {ethers.Contract} usdcContract
+     * @param {string} forwarderContractAddress
+     * @param {ethers.BigNumber} approvalAmount
+     * @param {number} tokenNonce
+     * @param {string} fromAddress
+     * @returns {Promise<{sigR: string, sigS: string, sigV: number}>}
+     */
+    async signUsdcApproval(path, usdcContract, forwarderContractAddress, approvalAmount, tokenNonce, fromAddress) {
+        const functionSignature = usdcContract.interface.encodeFunctionData(
+            'approve',
+            [forwarderContractAddress, approvalAmount],
+        );
+
+        // TODO: Make the domain parameters configurable in the request?
+        const domain = {
+            name: 'USD Coin (PoS)', // This is currently the same for testnet and mainnet
+            version: '1', // This is currently the same for testnet and mainnet
+            verifyingContract: CONFIG.USDC_CONTRACT_ADDRESS,
+            salt: ethers.utils.hexZeroPad(ethers.utils.hexlify(CONFIG.POLYGON_CHAIN_ID), 32),
+        };
+
+        const types = {
+            MetaTransaction: [
+                { name: 'nonce', type: 'uint256' },
+                { name: 'from', type: 'address' },
+                { name: 'functionSignature', type: 'bytes' },
+            ],
+        };
+
+        const message = {
+            nonce: tokenNonce,
+            from: fromAddress,
+            functionSignature,
+        };
+
+        const signature = await this.signTypedData(
+            path,
+            domain,
+            types,
+            message,
+        );
+
```

Comment:
This little extra-check of the signature was removed with the refactoring:

```suggestion

        const signerAddress = ethers.utils.verifyTypedData(domain, types, message, signature);
        if (signerAddress !== fromAddress) {
            reject(new Errors.CoreError('Failed to sign approval'));
            return;
        }

```

[But seems that you don't like it](https://github.com/nimiq/keyguard/pull/465#discussion_r1358902067), which is then also fine.

Reply:
Ha, you spotted that :D 

Yes, I think it's unnecessary now that we know it works.